src/vs/workbench/browser/parts/editor/editorPart.ts

```diff
@@ -364,26 +382,70 @@ export class EditorPart extends Part implements IEditorPart {
 			return; // we have not been created yet
 		}
 
+		if (this.isGroupMaximized(target)) {
+			if (arrangement === GroupsArrangement.MAXIMIZE && this.activeGroup === target) {
+				return; // already maximized
+			}
+			this.unmaximizeGroup();
+		}
+
 		switch (arrangement) {
 			case GroupsArrangement.EVEN:
 				this.gridWidget.distributeViewSizes();
 				break;
 			case GroupsArrangement.MAXIMIZE:
+				if (this.groups.length < 2) {
+					return; // need at least 2 groups to be maximized
+				}
+				this.gridWidget.setViewsVisible(false, target);
+				this._onDidChangeMaximizeGroup.fire({ group: target, maximized: true });
+				if (this.activeGroup !== target) {
+					this.activateGroup(target);
+				}
+				break;
+			case GroupsArrangement.EXPAND:
 				this.gridWidget.maximizeViewSize(target);
 				break;
-			case GroupsArrangement.TOGGLE:
-				if (this.isGroupMaximized(target)) {
-					this.arrangeGroups(GroupsArrangement.EVEN);
-				} else {
-					this.arrangeGroups(GroupsArrangement.MAXIMIZE);
-				}
+		}
+	}
 
-				break;
+	toggleGroupArrangement(): void {
+		if (this.isGroupMaximized(this.activeGroup)) {
+			this.unmaximizeGroup();
+		} else if (this.isGroupExpanded(this.activeGroup)) {
+			this.arrangeGroups(GroupsArrangement.EVEN);
+		} else {
+			this.arrangeGroups(GroupsArrangement.MAXIMIZE, this.activeGroup);
```

Comment:
I was expecting a setting for the user to pick the new behaviour but not have this enabled by default (yet, maybe in the future).

Reply:
You are right, a setting would be good. It could be something like `workbench.editor.tabDoubleClickAction`: [`maximizeGroup`, `expandGroup`]. This would allow future additions.
I however think the maximize should be the default from the start.

Reply:
A setting like that works for me. Not sure about the default :)

I think we have another candidate here for changing a `boolean` setting to an `enum` because we already have a setting: 

<img width="895" alt="image" src="https://github.com/microsoft/vscode/assets/900690/a0730f52-ac4f-4154-a749-4a0928d684ce">